prog/pwm/fancontrol.8
.TH FANCONTROL 8 "September 2009" "lm-sensors 3"
.SH NAME
fancontrol \- automated software based fan speed regulation

.SH SYNOPSIS
.B fancontrol [
.I --check
.B ] [
.I configfile
.B ]

.SH DESCRIPTION
\fBfancontrol\fP is a shell script for use with lm_sensors. It reads its
configuration from a file, then calculates fan speeds from temperatures and
sets the corresponding PWM outputs to the computed values.

.SH OPTIONS
.IP "--check"
Check configuration file end exit. This option can be used for verifying the
configuration before starting fancontrol as a service in the background.

.SH WARNING
Please be careful when using the fan control features of your mainboard, in
addition to the risk of burning your CPU, at higher temperatures there will be
a higher wearout of your other hardware components, too. So if you plan to use
these components in 50 years, \fBmaybe\fP you shouldn't use fancontrol
at all.  Also please keep in mind most fans aren't designed to be
powered by a PWMed voltage.

In practice it doesn't seem to be a major issue, the fans will get slightly
warmer, just be sure to have a temperature alarm and/or shutdown call, in case
some fan fails, because you probably won't hear it anymore ;)

.SH CONFIGURATION
For easy configuration, there's a script
named \fBpwmconfig\fP(8) which lets you interactively write your
configuration file for \fBfancontrol\fP. Alternatively you can write this
file yourself using the information from this manpage.

Since most of you are going to use \fBpwmconfig\fP(8) script, the config
file syntax will be discussed last. First I'm going to describe the various
variables available for changing \fBfancontrol\fP's behaviour:

.TP
.B ROOTDIR
An alternate directory of where to find device names. The default is
\fB/sys/class/hwmon\fP.
.TP
.B INTERVAL
This variable defines at which interval in seconds the main loop of
\fBfancontrol\fP will be executed
.TP
.B DEVPATH
Maps hwmon class devices to physical devices. This lets \fBfancontrol\fP
check that the configuration file is still up-to-date.
.TP
.B DEVNAME
Records hwmon class device names. This lets \fBfancontrol\fP check that
the configuration file is still up-to-date.
.TP
.B FCTEMPS
Maps PWM outputs to temperature sensors so \fBfancontrol\fP knows which
temperature sensors should be used for calculation of new values for
the corresponding PWM outputs. A prefix of ! can be used to execute a
command as a temperature sensor.

.TP
.B FCFANS
Records the association between a PWM output and a fan input.
Then \fBfancontrol\fP can check the fan speed and restart it if it
stops unexpectedly.
.TP
.B MINTEMP
The temperature below which the fan gets switched to minimum speed.
.TP
.B MAXTEMP
The temperature over which the fan gets switched to maximum speed.
.TP
.B MINSTART
Sets the minimum speed at which the fan begins spinning. You should
use a safe value to be sure it works, even when the fan gets old.
.TP
.B MINSTOP
The minimum speed at which the fan still spins. Use a safe value here,
too.
.TP
.B MINPWM
The PWM value to use when the temperature is below MINTEMP.
Typically, this will be either 0 if it is OK for the fan to plain
stop, or the same value as MINSTOP if you don't want the fan to
ever stop.
If this value isn't defined, it defaults to 0 (stopped fan).
.TP
.B MAXPWM
The PWM value to use when the temperature is over MAXTEMP.
If this value isn't defined, it defaults to 255 (full speed).
.TP
.B AVERAGE
How many last temperature readings are used to average the temperature.
It can be used to smoothen short temperature peaks.
If this value isn't defined, it defaults to 1 (no averaging).
.PP
The configuration file format is a bit strange:
.IP
.nf
VARIABLE=chip/pwmdev=value chip/pwmdev2=value2
VARIABLE2=...
.fi
.PP
Each variable has its own line. The variable name is followed by an equal sign
and the device=value pairs. These consist of the path to the pwm output for
which the value is valid, equal sign followed by the value and are separated
by a blank. Path can be absolute or relative (from /sys/bus/i2c/devices or
/sys/class/hwmon depending on the kernel version). Example:
.IP
MINTEMP=hwmon0/device/pwm1=40 hwmon0/device/pwm2=54
.PP
You have to play with the temperature values a bit to get happy. For initial
setup I recommend using the \fBpwmconfig\fP script. Small changes can be made by
editing the config file directly following the rules above.

Upon starting, fancontrol will make sure that all referenced devices
do exist and match what they were at configuration time, and that all
referenced sysfs files do exist. If not, it will quit immediately, upon
the assumption that the configuration file may be out-of-sync with the
loaded kernel drivers.

.SH THE ALGORITHM

\fBfancontrol\fP first reads its configuration, writes it to arrays and loops its
main function.  This function gets the temperatures and fanspeeds from
kernel driver files and calculates new speeds depending on temperature
changes, but only if the temp is between MINTEMP and MAXTEMP. After that, the
new values are written to the PWM outputs. Currently the speed increases
linearly with rising temperature. This way you won't hear your fans most
of the time at best.

.SH SEE ALSO
pwmconfig(8), sensors(1).

.SH AUTHOR
.PP
Marius Reiner <marius.reiner@hdev.de>


CHANGES
lm-sensors CHANGES file
-----------------------

git HEAD
  Add PWM sensor support

3.6.0 (2019-10-18)
  configs: Added a number of new configuration files
  fancontrol: AVERAGE env variable can be used to set the number of previous readings to average
  Makefile: The MACHINE variable has been renamed to ARCH
  sensord: Add an option -1/--oneline to print chip and adapter on the same line
  sensors: Fixed a stray comma bug in the JSON output
           Fixed Fahrenheit conversion with raw and JSON output
           Scale voltage and current values in the default output format
  sensors-detect: Add detection of AMD Family 17h, models 30h, 70h
                  Add detection of some AMD Family 15h models
                  Add detection of AMD Family 16h model 30h power sensors
                  Add detection of Hygon Family 18h thermal sensors
                  Add detection of Nuvoton NCT6797D
                  Add detection of Nuvoton NCT6798D
                  Add detection of Nuvoton NCT6112D/NCT6114D/NCT6116D
                  Fix printing CPU info on non-x86 arches
                  Fix printing lm_sensors version
                  Mark Fintek F75387SG/RG as supported by the f75375s driver

3.5.0 (2018-11-23)
  Fixed disappearance of certain hwmon chips with 4.19+ kernels
  Add the find-driver script for debugging
  Various documentation and man page improvements
  Fix various issues found by Coverity Scan
  Fix compilation with the musl C library
  Development version string now contains "+git" instead of "+SVN"
  Updated links in documentation to reflect the new home of lm_sensors
  sensors.1: Add reference to sensors-detect
             Document -j option (json output)
  sensors: Add support for json output
           Add support for power min, lcrit, min_alarm, lcrit_alarm
  sensors-detect: Fix systemd paths
                  Add detection of Fintek F81768
                  Only probe I/O ports on x86
                  Add detection of Nuvoton NCT6793D
                  Add detection of Microchip MCP9808
                  Mark F71868A as supported by the f71882fg driver
                  Mark F81768D as supported by the f71882fg driver
                  Mark F81866D as supported by the f71882fg driver
                  Add detection of various ITE chips
                  Add detection of Nuvoton NCT6795D
                  Add detection of DDR4 SPD
                  Add detection of ITE IT8987D
                  Add detection of AMD Family 17h temperature sensors
                  Add detection of AMD KERNCZ SMBus controller
                  Add detection of various Intel SMBus controllers
                  Add detection of Giantec GT30TS00
                  Add detection of ONS CAT34TS02C and CAT34TS04
                  Add detection of AMD Family 15h Model 60+ temperature sensors
                  Add detection of Nuvoton NCT6796D
                  Add detection of AMD Family 15h Model 70+ temperature sensors
  configs: Add sample configuration files.
  sensors.conf.default: Add hardwired inputs of NCT6795D
                        Add hardwired inputs of F71868A
                        Add hardwired NCT6796D inputs
  vt1211_pwm: replaced deprecated sub shell syntax
              run with bash instead of sh
  pwmconfig: replaced deprecated sub shell syntax
  fancontrol: replaced deprecated sub shell syntax
              save original pwm values
  fancontrol.8: replaced deprecated sub shell syntax
  libsensors: Add support for SENSORS_BUS_TYPE_SCSI
              Add support for power min, lcrit, min_alarm, lcrit_alarm

3.4.0 (2015-06-25)
  documentation: Update the note about libsensors license
  sensors.conf.5: Enhance the hysteresis documentation
  sensors.conf.default: Add support for NCT6779 and NCT6791
  libsensors: Get rid of arbitrary limit on per-type sensor count
              Add support for temperature min and critical min hysteresis
              Fix sparse warnings
  sensors: Add support for temperature min and critical min hysteresis
  fancontrol: Deal with moving hwmon attributes
              Fix shell error when FCFANS is not set
              Reduce memory consumption (#2391)
  healthd: Avoid using external commands
  sensors-detect: Add detection of ADC128D818
                  Add detection of AMD family 16h Kabini and Mullins
                  Add detection of ITE IT8620E and IT8623E
                  Add detection of TMP441, TMP442, LM95233, LM95234,
                  and LM95235
                  Add detection of NCT7802Y
                  Print kernel version and processor information
                  Add detection of NCT6792D 
                  Add reference to nct6683 driver
                  Show that NCT6102D/6104D/6106D are supported by nct6775 driver
                  Show NCT5573D as compatible to NCT6776F with same chip ID
                  Document support for EMC1402, EMC1404, and EMC1424
                  Detect new revisions of EMC14xx
                  Add detection of EMC1422
                  Document driver support for TMP431 and TMP432
                  Add detection of TMP400 and TMP435
                  Add detection of IT8786E
                  Add detection of IT8780F
                  Add detection of IT8731F and IT8732F
                  Add detection of Intel 5500/5520/X58
                  Loosen detection rules for TS3000/TSE2002
                  Add detection of TSE2004 and TS3001
                  Document driver support for NCT7802Y
                  Add detection of NCT7904D
                  Add detection of IT8790E
                  Fix graphics cards detection (#2386, #2392)
                  Skip more addresses on graphics card I2C bus (#2392)

3.3.5 "Happy Birthday Beddy" (2014-01-22)
  libsensors: Improve documentation of two functions
              Increase MAX_SENSORS_PER_TYPE to 33
  init: Fix EnvironmentFile in service files
        Drop syslog.target dependency from service files

prog/pwm/fancontrol.pl
#!/usr/bin/perl -wT
# $Id$
#
# Perl script for temperature dependent fan speed control.
#
# Copyright 2004 dean takemori <deant@hawaii.rr.com>
#
# This is a reimplementation in perl of Marius Reiner's bash script for
# fan speed control.  It has advantages in that it can daemonize itself
# and needn't spawn subprocesses for grep, sleep etc.  Much of the structure
# of the bash script is preserved to make mirroring changes easier, so
# this is seriously non-idiomatic perl but at the same time it should not
# be considered a a direct bash to perl translation.
#
# Usage: fancontrol [CONFIGFILE]
#
# For configuration instructions and warnings please see fancontrol.txt,
# which can be found in the doc/ directory or at the website mentioned
# elsewhere.
#
# This script is derived from Marius Reiner's bash version, so it is
# hereby placed under the GPL.
#
#    Copyright 2003 Marius Reiner <marius.reiner@hdev.de>
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#

use warnings;
use strict;
use IO::Handle;
use Getopt::Std;
use POSIX;

$ENV{PATH} = "/bin:/usr/bin";

##### Configuration #####
use constant DEBUG => 0;
use constant MAX   => 255;

use constant PIDFILE  => '/var/run/fancontrol.pid';
use constant CONFFILE => '/etc/fancontrol';
use constant LOGFILE  => '/var/log/fancontrol/fancontrol.log';
use constant ERRFILE  => '/var/log/fancontrol/fancontrol.err';
### End Configuration ###

our $interval;
our $pwmo;
our @afcpwm;
our @afctemp;
our @afcfan;
our @afcmaxtemp;
our @afcmintemp;
our @afcminstart;
our @afcminstop;

sub loadconfig($);
sub pwmdisable($);
sub pwmenable($);
sub restorefans();
sub calc(@);
sub UpdateFanSpeeds();
END { restorefans(); }

our $opt_d;
getopts('d');

my $config = shift;
if (defined($config))
   { loadconfig($config); }
else
   { loadconfig(CONFFILE); }

### Daemonize
if ( defined($opt_d) && ($opt_d == 1) )
   {
     my $pid = fork;
     POSIX::_exit(0) if $pid;

     unless (defined($pid))
       { die("Couldn't fork: $!"); }

     open(*STDERR, '>', ERRFILE);
     IO::Handle::autoflush(*STDERR);
     open(*STDOUT, '>>', LOGFILE);
     IO::Handle::autoflush(*STDOUT);

     unless (POSIX::setsid())
       { die("Couldn't open new session: $!"); }

   }

### Pidfile
if (open(FILE, ">" . PIDFILE))
   {
     print(FILE "$$\n");
     close(FILE);
   }
else
   { print(PIDFILE . ": $!\n"); }


### What kind of interface?
our $sysfs = 0;
our $dir = '/proc/sys/dev/sensors';
if (!(-d $dir))
   {
     if ($afcpwm[0] =~ m/^hwmon\d/)
       {
         $dir = '/sys/class/hwmon';
       }
     else
       {
         $dir = '/sys/bus/i2c/devices';
       }

     if (!(-d $dir))
       { die("No sensors found! (are the necessary modules loaded?) :
$!\n"); }
     else
       {
         $sysfs = 1;
       }
   }

### Trap signals
$SIG{TERM} = \&restorefans;
$SIG{HUP}  = \&restorefans;
$SIG{INT}  = \&restorefans;
$SIG{QUIT} = \&restorefans;

### Enable PWM
print("Enabling PWM on fans...\n");
my $fcvcount = 0;
while ($fcvcount < $#afcpwm+1)
   {
     $pwmo = $afcpwm[$fcvcount];
     unless (pwmenable($pwmo))
       { die("Error enabling PWM on $dir/$pwmo : $!\n"); }
     $fcvcount++;
   }

print("Starting automatic fan control...\n");

while(1)
   {
     UpdateFanSpeeds();
     sleep($interval);
   }

1;

################################################################
sub loadconfig($)
{
   my $file = shift;

   print("Loading configuration from $file ...\n");

   unless ( (-e $file) && (-r $file) )
     { die("Unable to read config file $file: $!"); }

   open(F, $file);

   our ($interval, $fctemps, $fcfans, $mintemp, $maxtemp, $minstart,
$minstop);
   while($_ = <F>)
     {
       if ($_ =~ /^\s+$/)                { next; }
       elsif ($_ =~ /^INTERVAL=\s*(.*)$/) { $interval = $1; next; }
       elsif ($_ =~ /^FCTEMPS=\s*(.*)$/)  { $fctemps = $1;  next; }
       elsif ($_ =~ /^FCFANS=\s*(.*)$/)   { $fcfans = $1;   next; }
       elsif ($_ =~ /^MINTEMP=\s*(.*)$/)  { $mintemp = $1;  next; }
       elsif ($_ =~ /^MAXTEMP=\s*(.*)$/)  { $maxtemp = $1;  next; }
       elsif ($_ =~ /^MINSTART=\s*(.*)$/) { $minstart = $1; next; }
       elsif ($_ =~ /^MINSTOP=\s*(.*)$/)  { $minstop = $1;  next; }
     }
   close(F);

   unless (defined($interval))
     { die("Some settings missing ..."); }

   print("\nCommon settings:\n");
   print("  INTERVAL=$interval\n");

   my $fcvcount = 0;
   foreach my $fcv (split(/\s+/, $fctemps))
     {
       ($afcpwm[$fcvcount], $afctemp[$fcvcount]) = split(/=/, $fcv);

       $fcfans   =~ s/^\S*=(\S+)\s*//;  $afcfan[$fcvcount]      = $1;
       $mintemp  =~ s/^\S*=(\S+)\s*//;  $afcmintemp[$fcvcount]  = $1;
       $maxtemp  =~ s/^\S*=(\S+)\s*//;  $afcmaxtemp[$fcvcount]  = $1;
       $minstart =~ s/^\S*=(\S+)\s*//;  $afcminstart[$fcvcount] = $1;
       $minstop  =~ s/^\S*=(\S+)\s*//;  $afcminstop[$fcvcount]  = $1;

       print("\nSettings for $afcpwm[$fcvcount]:\n");
       print("  Depends on $afctemp[$fcvcount]\n");
       print("  Controls $afcfan[$fcvcount]\n");
       print("  MINTEMP  = $afcmintemp[$fcvcount]\n");
       print("  MAXTEMP  = $afcmaxtemp[$fcvcount]\n");
       print("  MINSTART = $afcminstart[$fcvcount]\n");
       print("  MINSTOP  = $afcminstop[$fcvcount]\n");

       $fcvcount++;
     }
}


################################################################
sub pwmdisable($)
{
   my $p = shift;

   if ($sysfs == 1)
     {
       if (open(F, ">$dir/$p"))
         {
           print(F MAX . '\n');
           close(F);
         }
       else
         { die("$dir/$p : $!"); }

       my $enable = "$dir/${p}_enable";
       if (-f $enable)
         {
           if (open(F, ">$enable"))
             {
               print(F '0');
               close(F);
             }
           else
             { die("$enable : $!"); }
         }
     }
   else
     {
       if (open(F, ">$dir/$p"))
         {
           print(F MAX . ' 0');
           close(F);
         }
       else
         { die("$dir/$p : $!"); }
     }
   return(1);
}


#################################################################
sub pwmenable($)
{
   my $p = shift;

   if ($sysfs == 1)
     {
       my $enable = "$dir/${p}_enable";
       if (-f $enable)
         {
           if (open(F, ">$enable"))
             {
               print(F "1\n");
               close(F);
             }
           else
             { die("$enable : $!\n"); }
         }
     }
   else
     {
       if (open(F, ">$dir/$p"))
         {
           print(F MAX . " 1\n");
           close(F);
         }
       else
         { die("$dir/$p : $!\n"); }
     }
   return(1);
}


################################################################
sub restorefans()
{
   $SIG{TERM} = 'IGNORE';
   $SIG{HUP}  = 'IGNORE';
   $SIG{INT}  = 'IGNORE';
   $SIG{QUIT} = 'IGNORE';

   print("Aborting, restoring fans...\n");
   my $fcvcount = 0;

   while ( $fcvcount < $#afcpwm+1)
     {
       my $pwmo = $afcpwm[$fcvcount];
       &pwmdisable($afcpwm[$fcvcount]);
       $fcvcount++;
     }
   print("Verify fans have returned to full speed\n");
   POSIX:_exit(-1);
}


############################################################
sub UpdateFanSpeeds()
{
   my $fcvcount = 0;

   while ($fcvcount < $#afcpwm+1)
     {
       my $pwmo  = $afcpwm[$fcvcount];
       my $tsens = $afctemp[$fcvcount];
       my $fan   = $afcfan[$fcvcount];
       my $mint  = $afcmintemp[$fcvcount];
       my $maxt  = $afcmaxtemp[$fcvcount];
       my $minsa = $afcminstart[$fcvcount];
       my $minso = $afcminstop[$fcvcount];

       ### tval
       my $tval = 0;
       if (open(F, "$dir/$tsens"))
         {
           $tval = <F>;
           close(F);
         }
       else
         { die("Error reading temperature from $dir/$tsens"); }
       $tval =~ /([.\d]+)\s*$/;
       $tval = int($1);
       if ($sysfs == 1)
         { $tval /= 1000; }

       ### pwmpval
       my $pwmpval = 0;
       if (open(F, "$dir/$pwmo"))
         {
           $pwmpval = <F>;
           close(F);
         }
       else
         { die("Error reading PWM value from $dir/$pwmo"); }
       ($pwmpval) = split(/\s/, $pwmpval);

       ### fanval
       my $fanval = 0;
       if (open(F, "$dir/$fan"))
         {
           $fanval = <F>;
           close(F);
         }
       else
         { die("Error reading Fan value from $dir/$fan"); }
       $fanval =~ /(\d+)\s$/;
       $fanval = $1;

       ### DEBUG
       if (DEBUG == 1)
         {
           print("pwmo=$pwmo\n");
           print("tsens=$tsens\n");
           print("fan=$fan\n");
           print("mint=$mint\n");
           print("maxt=$maxt\n");
           print("minsa=$minsa\n");
           print("minso=$minso\n");
           print("tval=$tval\n");
           print("pwmpval=$pwmpval\n");
           print("fanval=$fanval\n");
           print("\n");
         }

       my $pwmval;
       if ($tval <= $mint)
         { $pwmval = 0; }
       elsif ($tval >= $maxt)
         { $pwmval = MAX; }
       else
         {
           $pwmval = eval ( ($tval - $mint) / ($maxt - $mint) );
           $pwmval *= (255 - $minso);
           $pwmval += $minso;
           $pwmval = int($pwmval);
           if ( ($pwmval == 0) || ($fanval == 0) )
             {
               if (open(F, ">$dir/$pwmo"))
                 {
                   print(F "$minsa\n");
                   close(F);
                 }
               else
                 { die("Error writing PWM value to $dir/$pwmo : $!\n"); }
               sleep 1;
             }
         }

       if (open(F, ">$dir/$pwmo"))
         {
           print(F "$pwmval\n");
           close(F);
         }
       else
         { die("Error writing PWM value to $dir/$pwmo : $!\n"); }

       if (DEBUG == 1)
         { print("new pwmval = $pwmval\n"); }

       $fcvcount++;
     }

}

1;


